### ISSUE TITLE
Is it possible to write a mutually recursive schema? 

### ISSUE THREAD

**Issue by Gnuxie** [sinclairzx81/typebox#1047]
I tried to use `Type.Index` on a recursive type ie using `Type.Index` on `This` as a hack. But this didn't seem to work with `StaticDecode` and all the indexed types were inferred as `never`

**Comment by Gnuxie** [comment]
That's so cool, I'll try this out as soon as I get back to my hobby project and I'll let you know what I think to it. Awesome work once again!

**Comment by sinclairzx81** [comment]
@Gnuxie All good :)

Will close off this issue for now, but let me know how you get on with the module / mutual recursion prototype. As mentioned, very keen to get feedback on it, so let me know your thoughts.

All the best!
S

**Comment by Gnuxie** [comment]
oki, so I had a go with it.

First thing to note is that I noticed that there's nothing on the type level to warn you if you create a typo while writing `ModuleRef`, I haven't figured out what will happen if i did that at runtime yet though. 

There might be a way around that if `ModuleRef` got provided via an argument in a callback, and from there you can probably make it so that the parameter of `ModuleRef` is typed to only allow the properties defined in the module. I'm not sure though and this is already does the job, and that might be considered noisy for little benefit. Thanks again, I'll come back as I keep playing with it.

```typescript
Module((ModuleRef) => ({
  MethodDescription: Type.Object(
    {
      slots: Type.Array(ModuleRef('SlotDescription')), // would error because the only property at the minute is `MethodDescription`.
      expression: ModuleRef("MethodExpression"),
    }
  ),
}));
```

**Comment by sinclairzx81** [comment]
@Gnuxie Hi, thanks for the feedback :)

> There might be a way around that if ModuleRef got provided via an argument in a callback, and from there you can probably make it so that the parameter of ModuleRef is typed to only allow the properties defined in the module. I'm not sure though and this is already does the job, and that might be considered noisy for little benefit. Thanks again, I'll come back as I keep playing with it.

Yeah, have explored this aspect quite a bit. The problem I ran into was that TS can't resolve properties keys within the context of a object literal. It can partially resolve within the context of a class however (as classes have access to the `this` context), but it still doesn't work quite right when used for establishing left side return types as the class signatures become cyclic when interior class members reference themselves....the issue is somewhat analogous to the following.

```typescript
// error: Return type annotation circularly references itself.
declare function X(): ReturnType<typeof X> 
```

The trick to making this work was to treat the ModuleRef string parameter as a kind of latent query to deference the type "after the fact". By making the string unaware of it's surroundings, you can mitigate cyclic referencing issues, and enable mutual recursion, but it does come at a cost of not being able to provide type hints of possible things to reference. I think it's ok tho, because if the ModuleRef string is incorrect, TypeBox can infer the type as TNever or just throw an error on calls to `.Import()`. 

TypeBox is working towards this design in the next revision, and I have been doing quite a bit of performance tuning around it.   A combination of deferred inference (as per ModuleRef (which will be Ref in the next release) and tail call optimized conditional types seems to yield really good scalable types. Will be looking for people to help me test that on existing codebases before the release, so if you're interested, I could use some help when things are a little closer to release.

At this stage, I had been hoping to get the next release out before the end of the year, but there are some other significant updates in the release (which I have been taking my time to get right), as well as exploring some very advanced new features I think people will like. With a bit of luck I'll get these updates out before the end of the year, and with that will come a push for TypeBox 1.0 in 2025 (fingers crossed)

Thanks again for the feedback!
S

**Comment by sinclairzx81** [comment]
@Gnuxie Hi,

Just a heads up, Module Types have just been published on 0.34.0. Documentation on the feature can be found at the following URL.

https://github.com/sinclairzx81/typebox#types-modules

This feature was pushed ahead inline with the new Syntax Types feature which supports TS module parsing.

https://github.com/sinclairzx81/typebox#syntax-types

And to provide future backend for the following project (with a plan to replace the TS compiler for code transformation)

https://sinclairzx81.github.io/typebox-workbench

The new features have taken quite a bit of research to achieve, so always open to feedback from the community on them. If you get a chance to use them and find any bugs, feel free to ping an issue (I expect these features will be a focus for the entirety of the 0.34.0 revision)

Cheers!
S

**Comment by Gnuxie** [comment]
Excellent, that's great news, thank you for the update. I probably won't get to play around again for some time but I'll get back to you if I get the opportunity!
